x/stakeibc/keeper/unbonding_records_get_host_zone_unbondings_msgs_test.go

```diff
@@ -58,7 +63,22 @@ func (s *KeeperTestSuite) SetupGetHostZoneUnbondingMsgs() GetHostZoneUnbondingMs
 			DelegationAmt: amtVal2,
 			Weight:        wgtVal2,
 		},
-	}
+	},
+	totalWeight: 5,
+	expectPass: true,
+}
+
+func (s *KeeperTestSuite) SetupGetHostZoneUnbondingMsgs(tc GetHostZoneUnbondingMsgsTestCase) (uint64, types.HostZone) {
```

Comment:
I like the approach of passing the TestCase as a parameter, so you don't have to use keeper functions to handle the different cases! 

However, I find having functions for each test case more digestible/maintainable than having the map of test cases. With the map, it's hard to diagnose and debug when each test is ran in a loop rather than individually. I know the map is often the standard in cosmos, but the other tests I've seen have been much smaller (and require less state) than some of these Stride tests.

What do you think / what was your rationale for using the map?

Reply:
There are some benefits of using table-driven tests for me:

- Easy to set up. For each change in the origin func, no need to set-up for all test funcs, just change the params
- Easy to look, other dev can see the difference of test-case from table
- Each test case run in its own context, if there're any err we can track which case threw them easily. Show example below
<img width="739" alt="Ảnh chụp Màn hình 2022-11-22 lúc 12 02 22" src="https://user-images.githubusercontent.com/72878483/203226421-f108f9dd-6d7d-4aac-8ce3-8607a7ec5eab.png">
<img width="1042" alt="Ảnh chụp Màn hình 2022-11-22 lúc 12 02 57" src="https://user-images.githubusercontent.com/72878483/203226428-38baf990-65db-49cf-bec0-3c0ca0e1765f.png">

Reply:
I think it brings a couple downsides with it though:
* While it saves ~1-2 lines of code per test by avoiding repeating the setup call, it adds a lot of duplicate code by repeating the struct definitions across each test (e.g. `epochUnbondingRecords` and `validators` are duplicated across each test). This makes it slightly harder to see what's being tested in each test.
* It makes it harder to debug because you have to run the full array of tests each time you make a change (i.e. there's no easy way to test one specific test case through an IDE). With separate functions, you can easily run each test individually (see the play button and "run test" option in the image below)
 ![Screen Shot 2022-11-22 at 11 47 20 AM](https://user-images.githubusercontent.com/19301583/203385148-ae54b9c7-6338-4d8b-9fbb-7159fe25cb13.png)
* It prevents the use of custom logic in each test (irrelevant for these tests, but we have other tests in the repo surrounding ICAs where we want to do things like close a channel in the middle of a test)
* It's also worth mentioning, that with the function setup, each test also runs in parallel in it's own context.

---

I think a good middle ground between what was originally in the code base and what you provided, would be to pass the test case as an argument, and provide functions for if there's any checks that seem to be repeated. 

that way the error tests will look something like this:

```go
var defaultUnbondingTestCase = ...

func (s *KeeperTestSuite) SetupGetHostZoneUnbondingMsgs(tc GetHostZoneUnbondingMsgsTestCase) (uint64, stakeibc.HostZone) {
   ...
}

func (s *KeeperTestSuite) TestGetHostZoneUnbondingMsgs_WrongChainId() {
   invalidState := defaultUnbondingTestCase
   invalidState.ChainId := "nonExistentChainId"

   resp, err := s.App.StakeibcKeeper.SomeFunctionBeingTested(s.Ctx(), invalidState)
   s.Require().EqualError(...)
}
```

This, in my opinion, makes it much easier to see what's going on in the test and makes the test file more modular and digestible overall.

And if you saw a lot of repeated logic, you could also always wrap the main function call + validation in it's own function so it looks like this: 
```go
func (s *KeeperTestSuite) VerifyTestCase(tc GetHostZoneUnbondingMsgsTestCase, success bool, err error) {
   resp, err := s.App.StakeibcKeeper.SomeFunctionBeingTested(s.Ctx(), invalidState)
   if success {
        ...
   }
   if err != nil {
        ...
   }
}

func (s *KeeperTestSuite) TestGetHostZoneUnbondingMsgs_WrongChainId() {
   invalidState := defaultUnbondingTestCase
   invalidState.ChainId := "nonExistentChainId"

   VerifyTestCase(invalidState)
}

```

What do you think?

Reply:
Agree with the idea of ​​using 1 object and then modifying it for each test case instead of using multiple.

Reply:
The design LGTM